### ISSUE TITLE
ActiveRecord::VERSION

### ISSUE THREAD

**Issue by obrie** [pluginaweek/state_machine#51]
**germaninthetown** opened this issue

Before all, thank you Aaron for this great gem. It&rsquo;s amazing. 

However I got an error while using it in my rails 2.1 application (yes, I know, it&rsquo;s old).

$ rails -v
Rails 2.1.0

$ ruby -v
ruby 1.8.6 (2008-08-08 patchlevel 286) [i686-linux]

I have a model Cart 

class Cart < ActiveRecord::Base

  ...

  state_machine :fullfilment_status, :initial => :incomplete do
    state :waiting_for_cutout_processing, :value => "Waiting for Manual Cutout Processing"
    state :processing,                    :value => "Cutouts being Processed"
    state :waiting_for_customer_approval, :value => "Waiting for Customer Approval"
    state :ready_to_print,                :value => "Ready to Print"
    state :being_printed,                 :value => "Being Printed"
    state :shipped,                       :value => "Shipped"
    state :cancelled,                     :value => "Cancelled"
    state :incomplete,                    :value => "incomplete"

```
state :reclamation,                   :value => "Reclamation - being reprinted"

event :purchase do               
  transition :incomplete => :waiting_for_cutout_processing     
end

event :wait_for_approval do
  transition :waiting_for_cutout_processing => :waiting_for_customer_approval
end

event :process do 
  transition :waiting_for_cutout_processing => :processing, :waiting_for_customer_approval => :processing
end

event :cancel do 
  transition all => :cancelled
end

event :ship do
  transition :processed => :shipped
end

event :print do
  transition :ready_to_print => :processing 
end

event :reclaim do
  transition :shipped => :being_printed
end
```

  end

  ...
end

but when I got to the ./script/console I got the error:

=> Cart.last
NameError: uninitialized constant ActiveRecord::VERSION
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:278:in `load_missing_constant&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:467:in`const_missing&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/integrations/active_record.rb:468:in `define_scope&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/integrations/active_record.rb:446:in`create_with_scope&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:1523:in `send&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:1523:in`define_scopes&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:1520:in `each&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:1520:in`define_scopes&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:1519:in `each&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:1519:in`define_scopes&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:445:in `initialize&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:332:in`new&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine/machine.rb:332:in `find_or_create&rsquo;
    from /opt/ruby-enterprise-1.8.6-20090201/lib/ruby/gems/1.8/gems/state_machine-0.9.0/lib/state_machine.rb:378:in`state_machine&rsquo;
    from /home/german/projects/ltlprints/app/models/cart.rb:11
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:215:in `load_without_new_constant_marking&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:215:in`load_file&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:354:in `new_constants_in&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:214:in`load_file&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:95:in `require_or_load&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:260:in`load_missing_constant&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:467:in `const_missing&rsquo;
    from /home/german/.gem/ruby/1.8/gems/activesupport-2.1.0/lib/active_support/dependencies.rb:479:in`const_missing&rsquo;
    from (irb):1>>

I tried to find the error and finally got the code working by adding the following line

require &rsquo;active_record/version&rsquo;

to the method self.extended (line 325) in /state_machine-0.9.0/lib/state_machine/integrations/active_record.rb file. 

I&rsquo;m not quite sure is this error appearing for me only, or I just did something wrong elsewhere in the code, but I think it would be great, if you&rsquo;ll add this line to your code so other developers wouldn&rsquo;t notice this error in future. How do you think?

Thanks,
Dmitrii Samoilov.

[original LH ticket](http://pluginaweek.lighthouseapp.com/projects/13288/tickets/41)

 This ticket has 0 attachment(s).

**Comment by obrie** [comment]
Hey Dmitrii,

Thanks for reporting this!  It turns out the tests that run ActiveRecord load active_record/version, causing this failure to be missed when testing against 2.1.0.

Will have a fix available soon.

**Comment by obrie** [comment]
(from [e9f8f0e4a22c9adae86e57734cad4574f6487bb6]) Fix ActiveRecord 2.0.0 - 2.2.3 integrations failing if version info isn&rsquo;t already loaded [#41 state:resolved]
http://github.com/pluginaweek/state_machine/commit/e9f8f0e4a22c9adae86e57734cad4574f6487bb6
